iina/LogWindowController.swift

```diff
@@ -0,0 +1,135 @@
+  }
+
+  @IBOutlet weak var logTableView: NSTableView!
+  @IBOutlet var logArrayController: NSArrayController!
+  @IBOutlet weak var subsystemPopUpButton: NSPopUpButton!
+  @IBOutlet weak var levelPopUpButton: NSPopUpButton!
+
+  @objc dynamic var logs: [Logger.Log] {
+    Logger.logs
+  }
+  @objc dynamic var predicate = NSPredicate(value: true)
+
+  override func windowDidLoad() {
+    super.windowDidLoad()
+
+    logTableView.sizeLastColumnToFit()
+    let tableViewMenu = NSMenu()
+    tableViewMenu.addItem(withTitle: "Copy", action: #selector(menuCopy), keyEquivalent: "")
+    logTableView.menu = tableViewMenu
+
+    levelPopUpButton.menu?.items.forEach {
+      $0.image = LogWindowController.indicatorIcon(withColor: colorMap[$0.tag]!)
+    }
+    levelPopUpButton.selectItem(withTag: Logger.Level.preferred.rawValue)
+    subsystemPopUpButton.menu!.delegate = self
+  }
+
+  fileprivate static func indicatorIcon(withColor color: NSColor) -> NSImage {
+    if let cached = circleDict[color] {
+      return cached
+    }
+    let image = NSImage(size: NSMakeSize(kIconSize, kIconSize), flipped: false) { rect in
+      let inset = NSInsetRect(rect, kBorderWidth / 2 + rect.size.width * 0.25, kBorderWidth / 2 + rect.size.height * 0.25)
+      let path = NSBezierPath.init(ovalIn: inset)
+      path.lineWidth = kBorderWidth
+
+      let fractionOfBlendedColor: CGFloat
+      if #available(macOS 10.14, *) {
+        fractionOfBlendedColor = (NSApp.appearance?.isDark ?? false) ? 0.15 : 0.3
+      } else {
+        fractionOfBlendedColor = 0.15
+      };
+      let borderColor = color.blended(withFraction: fractionOfBlendedColor, of: .controlTextColor)
+
+      borderColor?.setStroke()
+      path.stroke()
+
+      color.setFill()
+      path.fill()
+
+      return true
+    }
+    circleDict[color] = image
+    return image
+  }
+
+  // MARK: - NSMenuDelegate
+
+  func menuNeedsUpdate(_ menu: NSMenu) {
+    Logger.subsystems.forEach {
+      if !$0.added {
+        menu.addItem(withTitle: $0.rawValue)
+        $0.added = true
+      }
+    }
+  }
+
+  private func updatePredicate() {
+    var subsystemPredicate = NSPredicate(value: true)
+    if subsystemPopUpButton.indexOfSelectedItem != 0 {
+      subsystemPredicate = NSPredicate(format: "subsystem = %@", subsystemPopUpButton.titleOfSelectedItem!)
+    }
+    let levelPredicate = NSPredicate(format: "level >= %d", levelPopUpButton.selectedTag())
+    predicate = NSCompoundPredicate(andPredicateWithSubpredicates: [subsystemPredicate, levelPredicate])
+  }
+
+  @IBAction func subsystemUpdated(_ sender: Any) {
+    updatePredicate()
+  }
+
+  @IBAction func save(_ sender: Any) {
+    Utility.quickSavePanel(title: "Log", filename: "iina.log", sheetWindow: window) { url in
+      let logs = (self.logArrayController.content as! [Logger.Log]).map { $0.logString }.joined()
+      try? logs.write(to: url, atomically: true, encoding: .utf8)
+    }
+  }
+
+  // Menu actions
```

Comment:
Should be: // MARK: - Menu actions
to take advantage of Xcode's source navigation feature as was done for the section comment above.

Reply:
Confirmed fixed.